Lib/idlelib/pyshell.py

```diff
@@ -1304,12 +1304,21 @@ def recall(self, s, event):
             self.text.see("insert")
             self.text.undo_block_stop()
 
+    _last_newline_re = re.compile(r"[ \t]*(\n[ \t]*)?$")
+    @classmethod
+    def _remove_last_newline_and_surrounding_whitespace(cls, line):
+        "Strip off last newline and surrounding whitespace."
+        # Add an extra newline at the end due to the way the re module
+        # treats the '$' symbol: "Matches the end of the string or just
+        # before the newline at the end of the string, ..."
+        return cls._last_newline_re.sub("", line + '\n')
```

Comment:
Consider using `\Z` instead of `$`.  It always means the very end of the string, irrespective of re.MULTILINE.

Reply:
The `+ '\n'` workaround looks also OK, but with \Z you won't need a long explanation.

BTW, I'm not sure the docstring is 100% accurate.  The regexp _optionally_ removes 1 newline but the docstring sounds like it'll always remove 1 newline?

Reply:
Thanks @cben, I'd forgotten about `\Z`! I've simplified this code using it.

Reply:
> BTW, I'm not sure the docstring is 100% accurate. The regexp optionally removes 1 newline but the docstring sounds like it'll always remove 1 newline?

The `+ '\n'` workaround isn't necessarily general, but it works specifically for this use-case. (Anyways, no longer needed thanks to `\Z` :))

Reply:
Thank you Beni.  Very helpful.